File: tools/frida_roombtn.py
"""Anexa o Frida no CLIENTE rodando e carrega frida_roombtn.js — loga a criacao dos botoes ao
entrar numa sala, p/ LOCALIZAR a funcao da tela do game room (em gamemp.dll) a patchar.

Uso:
  1) Abra o jogo (launcher) e fique no LOBBY (lista de salas).
  2) python tools/frida_roombtn.py            # acha o cliente por nome
     (ou)  python tools/frida_roombtn.py 1234  # PID do cliente
  3) ENTRE numa sala (game room). Os 6 botoes serao logados aqui com 'caller=gamemp.dll+0x....'.
  4) Me mande a saida (as linhas SetText/CTOR/SetPos/SetSize com os callers).

Precisa: pip install frida   (o projeto ja usa frida em rakion-work).
"""
import frida, sys, time, os

JS = os.path.join(os.path.dirname(os.path.abspath(__file__)), "frida_roombtn.js")
CLIENT_NAMES = ("rakion.exe", "rakion.bin", "Rakion.exe")

def on_message(msg, data):
    t = msg.get("type")
    if t == "send":
        print(msg.get("payload"), flush=True)
    elif t == "error":
        print("JS-ERROR:", msg.get("description"), flush=True)
    else:
        print(msg, flush=True)

def find_client_pid():
    # frida 16: enumerate_processes saiu do modulo -> via device local
    dev = frida.get_local_device()
    for p in dev.enumerate_processes():
        if p.name in CLIENT_NAMES or "rakion" in p.name.lower():
            return p.pid, p.name
    return None, None

def main():
    if len(sys.argv) > 1:
        pid = int(sys.argv[1]); name = "(pid)"
    else:
        pid, name = find_client_pid()
        if not pid:
            print("[erro] cliente Rakion nao encontrado. Abra o jogo primeiro (ou passe o PID).")
            return
    print("[driver] anexando no %s pid=%d ..." % (name, pid), flush=True)
    session = frida.attach(pid)
    with open(JS, "r", encoding="utf-8") as f:
        src = f.read()
    script = session.create_script(src)
    script.on("message", on_message)
    script.load()
    print("[driver] hooks ativos. ENTRE numa SALA no jogo. (capturando ~30min; Ctrl+C p/ sair)", flush=True)
    try:
        time.sleep(1800)
    except KeyboardInterrupt:
        pass
    print("[driver] fim.", flush=True)

if __name__ == "__main__":
    main()

File: tools/frida_roombtn.js
'use strict';
// Hook das criacoes de botao da UI (uitoolkit.dll csButton) p/ LOCALIZAR a tela do GAME ROOM em
// gamemp.dll. Ao entrar numa sala, loga cada botao: texto, posicao e o CALLER (modulo+offset = a
// funcao que cria a tela). Com isso eu cravo onde injetar o botao "Adicionar Bot".
// x86 __thiscall: this=ECX; args na pilha a partir de [esp+4]; retaddr = this.returnAddress.

function out(m) { send(m); }

function callerOf(ret) {
  try {
    var m = Process.findModuleByAddress(ret);
    if (!m) return ret.toString();
    return m.name + '+0x' + ret.sub(m.base).toString(16);
  } catch (e) { return '?'; }
}

function hook(mangled, label, onEnter) {
  try {
    var addr = Module.getExportByName('uitoolkit.dll', mangled);
    Interceptor.attach(addr, { onEnter: onEnter });
    out('[hook] ' + label + ' @ ' + addr);
  } catch (e) { out('[ERRO] ' + label + ': ' + e.message); }
}

// csButton::SetText(char* text)  -> this=ECX, text=[esp+4]
hook('?SetText@csButton@@UAEXPBD@Z', 'csButton::SetText', function () {
  var self = this.context.ecx;
  var p = this.context.esp.add(4).readPointer();
  var txt = p.isNull() ? '' : p.readCString();
  out('SetText  btn=' + self + ' text="' + txt + '"  caller=' + callerOf(this.returnAddress));
});

// csButton::ctor(csComponent* parent, int a, int b, frameStyle) -> this=ECX, parent=[esp+4], a=[esp+8], b=[esp+12]
hook('??0csButton@@QAE@PAVcsComponent@@HHW4csComponentFrameStyle@@@Z', 'csButton::ctor', function () {
  var self = this.context.ecx;
  var parent = this.context.esp.add(4).readPointer();
  var a = this.context.esp.add(8).readS32();
  var b = this.context.esp.add(12).readS32();
  out('CTOR     btn=' + self + ' parent=' + parent + ' id?=' + a + '/' + b + '  caller=' + callerOf(this.returnAddress));
});

// csComponent::SetPos(long x, long y) -> this=ECX, x=[esp+4], y=[esp+8]
hook('?SetPos@csComponent@@QAEXJJ@Z', 'csComponent::SetPos', function () {
  var self = this.context.ecx;
  var x = this.context.esp.add(4).readS32();
  var y = this.context.esp.add(8).readS32();
  out('SetPos   btn=' + self + ' x=' + x + ' y=' + y + '  caller=' + callerOf(this.returnAddress));
});

// csComponent::SetSize(long w, long h)
hook('?SetSize@csComponent@@QAEXJJ@Z', 'csComponent::SetSize', function () {
  var self = this.context.ecx;
  var w = this.context.esp.add(4).readS32();
  var h = this.context.esp.add(8).readS32();
  out('SetSize  btn=' + self + ' w=' + w + ' h=' + h + '  caller=' + callerOf(this.returnAddress));
});

out('[roombtn] hooks armados. ENTRE numa SALA (game room) — os 6 botoes (Previous/Messenger/Invite/');
out('[roombtn] Game setting/Team change/Start) serao logados com o caller = a funcao da tela a patchar.');
